### PR TITLE
chore: Test all coroutine patterns and refactor op call

### PR DESCRIPTION
**Pull request by tssweeney**
I am working on another feature and needed to use the `Op.call` pattern. I realized that the typing for this was not exactly correct AND we did not use the same code as the wrapper logic.

So, this PR:
1. Adds a bunch of tests to ensure we do the correct thing for coroutines
2. Shares as much code as possible between `call`, `async wrapper` and `wrapper`
3. Improves typing in various locations

### CODE REVIEW

File: weave/trace/weave_client.py [hunk 1]
```diff
@@ -715,6 +715,8 @@ def finish_call(
         *,
         op: Optional[Op] = None,
     ) -> None:
+        ended_at = datetime.datetime.now(tz=datetime.timezone.utc)
+        call.ended_at = ended_at
```
Comment (tssweeney) [inline weave/trace/weave_client.py:719]: small bug i noticed - we don't update ended at!

File: weave/trace/op.py [hunk 2]
```diff
@@ -453,51 +560,17 @@ def create_wrapper(func: Callable) -> Op:
...
-                            raise
-                        log_once(
-                            logger.error,
-                            ASYNC_CALL_CREATE_MSG.format(traceback.format_exc()),
-                        )
-                        return await func(*args, **kwargs)
-                    res, _ = await _execute_call(wrapper, call, *args, **kwargs)  # type: ignore
+                    res, _ = await do_call_async(
+                        cast(Op, wrapper), *args, __should_raise=True, **kwargs
+                    )
```
Comment (andrewtruong) [inline weave/trace/op.py:565]: am I correct in reading that you extracted this into `do_call_async`?  Is anything else happening here?

File: weave/trace/op.py [hunk 3]
```diff
@@ -453,51 +560,17 @@ def create_wrapper(func: Callable) -> Op:
...
-                        if get_raise_on_captured_errors():
-                            raise
-                        log_once(
-                            logger.error, CALL_CREATE_MSG.format(traceback.format_exc())
-                        )
-                        return func(*args, **kwargs)
-                    res, _ = _execute_call(wrapper, call, *args, **kwargs)  # type: ignore
+                    res, _ = do_call(
+                        cast(Op, wrapper), *args, __should_raise=True, **kwargs
+                    )
```
Comment (andrewtruong) [inline weave/trace/op.py:573]: same question here

File: weave/trace/op.py (hunk 2 above)
Comment (tssweeney) [reply weave/trace/op.py:565]: yes, no logical changes. But the key now is that `def call` can dispatch to each of these

File: weave/trace/op.py [hunk 4]
```diff
@@ -332,8 +336,111 @@ def add(a: int, b: int) -> int:
...
+    return Call(
+        _op_name="",
+        trace_id="",
+        project_id="",
+        parent_id=None,
+        inputs={},
+    )
+
+
+def do_call(
```
Comment (andrewtruong) [inline weave/trace/op.py:362]: These are private, so should be underscored

File: weave/trace/op.py [hunk 5]
```diff
@@ -332,8 +336,111 @@ def add(a: int, b: int) -> int:
...
+            if inspect.iscoroutine(execute_result):
+                raise Exception(
+                    "Internal error: Expected `_execute_call` to return a sync result"
+                )
+            execute_result = cast(tuple[Any, "Call"], execute_result)
+            res, call = execute_result
+    return res, call
+
+
+async def do_call_async(
```
Comment (andrewtruong) [inline weave/trace/op.py:403]: These are private, so should be underscored

Comment (tssweeney) [reply weave/trace/op.py]: ack - done

File: tests/trace/test_op_decorator_behaviour.py [hunk 6]
```diff
@@ -141,7 +141,7 @@ def test_sync_method_call(client, weave_obj, py_obj):
         weave_obj_method2 = weave_obj_method_ref.get()
 
     with pytest.raises(errors.OpCallError):
-        res2, call2 = py_obj.amethod.call(1)
+        res2, call2 = py_obj.method.call(1)
```
Comment (tssweeney) [inline tests/trace/test_op_decorator_behaviour.py:144]: this was a small but in the test suite actually. This should have been testing `method` not `amethod`

File: tests/trace/test_op_coroutines.py [hunk 7]
```diff
@@ -0,0 +1,220 @@
+import asyncio
+from typing import Coroutine
+
+import pytest
+
+import weave
+from weave.trace.weave_client import Call
+
+
+def test_non_async_non_coro(client):
```
Comment (andrewtruong) [inline tests/trace/test_op_coroutines.py:10]: I would just call this `test_sync`

File: tests/trace/test_op_coroutines.py [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+        return 1
+
+    res = non_async_non_coro()
+    assert res == 1
+    res, call = non_async_non_coro.call()
+    assert isinstance(call, Call)
+    assert res == 1
+
+
+def test_non_async_non_coro_method(client):
```
Comment (andrewtruong) [inline tests/trace/test_op_coroutines.py:22]: `test_sync_method`

File: tests/trace/test_op_coroutines.py [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+    test_inst = TestClass()
+    res = test_inst.non_async_non_coro()
+    assert res == 1
+    res, call = test_inst.non_async_non_coro.call(test_inst)
+    assert isinstance(call, Call)
+    assert res == 1
+
+
+@pytest.mark.asyncio
+async def test_non_async_coro(client):
```
Comment (andrewtruong) [inline tests/trace/test_op_coroutines.py:37]: `test_sync_in_coroutine`

File: tests/trace/test_op_coroutines.py [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+    assert isinstance(res, Coroutine)
+    assert await res == 1
+    res, call = non_async_coro.call()
+    assert isinstance(call, Call)
+    assert isinstance(res, Coroutine)
+    assert await res == 1
+
+
+@pytest.mark.asyncio
+async def test_non_async_coro_method(client):
```
Comment (andrewtruong) [inline tests/trace/test_op_coroutines.py]: `test_sync_method_in_coroutine`

File: tests/trace/test_op_coroutines.py (hunk 8 above)
Comment (tssweeney) [reply tests/trace/test_op_coroutines.py]: done

File: tests/trace/test_op_coroutines.py (hunk 9 above)
Comment (tssweeney) [reply tests/trace/test_op_coroutines.py]: opposite, but yeah - changed

File: tests/trace/test_op_coroutines.py [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+    res, call = test_inst.non_async_coro.call(test_inst)
+    assert isinstance(call, Call)
+    assert isinstance(res, Coroutine)
+    assert await res == 1
+
+
+@pytest.mark.asyncio
+async def test_async_coro(client):
+    @weave.op()
+    async def async_coro():
```
Comment (andrewtruong) [inline tests/trace/test_op_coroutines.py:71]: when would you write code like this?

File: tests/trace/test_op_coroutines.py (hunk 10 above)
Comment (tssweeney) [reply tests/trace/test_op_coroutines.py]: changed everything to `test_[Op_type]_[return_type]`

File: tests/trace/test_op_coroutines.py (hunk 11 above)
Comment (tssweeney) [reply tests/trace/test_op_coroutines.py:71]: not normally, but better to test

File: tests/trace/test_op_coroutines.py [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+    test_inst = TestClass()
+    res = test_inst.async_non_coro()
+    assert isinstance(res, Coroutine)
+    assert await res == 1
+    res, call = await test_inst.async_non_coro.call(test_inst)
+    assert isinstance(call, Call)
+    assert res == 1
+
+
+def test_non_async_with_exception(client):
```
Comment (andrewtruong) [inline tests/trace/test_op_coroutines.py]: yeah basically anything that is marked "non_async" should just be called "sync"

File: tests/trace/test_op_coroutines.py [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+        def non_async_with_exception(self):
+            raise ValueError("test")
+
+    test_inst = TestClass()
+    with pytest.raises(ValueError, match="test"):
+        test_inst.non_async_with_exception()
+    res, call = test_inst.non_async_with_exception.call(test_inst)
+    assert isinstance(call, Call)
+    assert call.exception is not None
+    assert res == None
```
Comment (andrewtruong) [inline tests/trace/test_op_coroutines.py:190]: all `None` checks should be `res is None`

Comment (tssweeney) [reply tests/trace/test_op_coroutines.py:190]: sure, changed

**Review by andrewtruong — APPROVED** [review]

**Pull request merged**